rslib/src/scheduler/filtered/custom_study.rs

```diff
@@ -0,0 +1,226 @@
+
+use std::cmp::Ordering;
+
+use super::FilteredDeckForUpdate;
+use crate::{
+    backend_proto::{
+        self as pb,
+        custom_study_request::{cram::CramKind, Cram, Value as CustomStudyValue},
+    },
+    decks::{FilteredDeck, FilteredSearchOrder, FilteredSearchTerm},
+    error::{CustomStudyError, FilteredDeckError},
+    prelude::*,
+    search::{
+        concatenate_searches, BoolSeparator, Node, PropertyKind, RatingKind, SearchNode, StateKind,
+    },
+    text::escape_anki_wildcards_for_search_node,
+};
+use itertools::Itertools;
+
+impl Collection {
+    pub fn custom_study(&mut self, input: pb::CustomStudyRequest) -> Result<OpOutput<()>> {
+        self.transact(Op::CreateCustomStudy, |col| col.custom_study_inner(input))
+    }
+}
+
+impl Collection {
+    fn custom_study_inner(&mut self, input: pb::CustomStudyRequest) -> Result<()> {
+        let current_deck = self.get_current_deck()?;
+
+        match input
+            .value
+            .ok_or_else(|| AnkiError::InvalidInput("missing oneof value".to_string()))?
+        {
+            CustomStudyValue::NewLimitDelta(delta) => {
+                let today = self.current_due_day(0)?;
+                self.extend_limits(today, self.usn()?, current_deck.id, delta, 0)
+            }
+            CustomStudyValue::ReviewLimitDelta(delta) => {
+                let today = self.current_due_day(0)?;
+                self.extend_limits(today, self.usn()?, current_deck.id, 0, delta)
+            }
+            CustomStudyValue::ForgotDays(days) => {
+                self.create_custom_study_deck(forgot_config(current_deck.human_name(), days))
+            }
+            CustomStudyValue::ReviewAheadDays(days) => {
+                self.create_custom_study_deck(ahead_config(current_deck.human_name(), days))
+            }
+            CustomStudyValue::PreviewDays(days) => {
+                self.create_custom_study_deck(preview_config(current_deck.human_name(), days))
+            }
+            CustomStudyValue::Cram(cram) => {
+                self.create_custom_study_deck(cram_config(current_deck.human_name(), cram)?)
+            }
+        }
+    }
+
+    /// Reuse existing one or create new one if missing.
+    /// Guaranteed to be a filtered deck.
+    fn create_custom_study_deck(&mut self, config: FilteredDeck) -> Result<()> {
+        let mut id = DeckId(0);
+        let human_name = self.tr.custom_study_custom_study_session().to_string();
+
+        if let Some(did) = self.get_deck_id(&human_name)? {
+            if !self
+                .get_deck(did)?
+                .ok_or(AnkiError::NotFound)?
+                .is_filtered()
+            {
+                return Err(CustomStudyError::ExistingDeck.into());
+            }
+            id = did;
+        }
+
+        self.add_or_update_filtered_deck_inner(FilteredDeckForUpdate {
+            id,
+            human_name,
+            config,
+        })
+        .map(|_| ())
```

Comment:
Very minor, and may come down to personal preference, but I find it can make things a little clearer if longer arguments are first aliased to a temporary variable, eg

```rust
        let deck = FilteredDeckForUpdate {
            id,
            human_name,
            config,
        };
        self.add_or_update_filtered_deck_inner(deck)
            .map(|_| ())
```